lindi/LindiStagingStore/LindiStagingStore.py

```diff
@@ -0,0 +1,267 @@
+        # would be dangerous if the file was part of a consolidated file.
+        return self._base_store.__delitem__(key)
+
+    def __iter__(self):
+        return self._base_store.__iter__()
+
+    def __len__(self):
+        return self._base_store.__len__()
+
+    # These methods are overridden from BaseStore
+    def is_readable(self):
+        return True
+
+    def is_writeable(self):
+        return True
+
+    def is_listable(self):
+        return True
+
+    def is_erasable(self):
+        return False
+
+    def _set_ref_reference(self, key: str, filename: str, offset: int, size: int):
+        rfs = self._base_store.rfs
+        if 'refs' not in rfs:
+            # this shouldn't happen, but we'll be defensive
+            rfs['refs'] = {}
+        rfs['refs'][key] = [
+            filename,
+            offset,
+            size
+        ]
+
+    def upload(
+        self,
+        *,
+        on_upload_blob: UploadFileFunc,
+        on_upload_main: UploadFileFunc,
+        consolidate_chunks: bool = True
+    ):
+        """
+        Consolidate the chunks in the staging area, upload them to a storage
+        system, updating the references in the base store, and then upload the
+        updated reference file system .json file.
+
+        Parameters
+        ----------
+        on_upload_blob : StoreFileFunc
+            A function that takes a string path to a blob file, uploads or copies it
+            somewhere, and returns a string URL (or local path).
+        on_upload_main : StoreFileFunc
+            A function that takes a string path to the main .json file, stores
+            it somewhere, and returns a string URL (or local path).
+        consolidate_chunks : bool
+            If True (the default), consolidate the chunks in the staging area
+            before uploading.
+
+        Returns
+        -------
+        str
+            The URL (or local path) of the uploaded reference file system .json
+            file.
+        """
+        if consolidate_chunks:
+            self.consolidate_chunks()
+        rfs = self._base_store.rfs
+        rfs = json.loads(json.dumps(rfs))  # deep copy
+        LindiReferenceFileSystemStore.replace_meta_file_contents_with_dicts(rfs)
+        blob_mapping = _upload_directory_of_blobs(self._staging_area.directory, on_upload_blob=on_upload_blob)
+        for k, v in rfs['refs'].items():
+            if isinstance(v, list) and len(v) == 3:
+                url1 = v[0]
+                if url1.startswith(self._staging_area.directory + '/'):
+                    url2 = blob_mapping.get(url1, None)
+                    if url2 is None:
+                        raise ValueError(f"Could not find url in blob mapping: {url1}")
+                    rfs['refs'][k][0] = url2
+        with tempfile.TemporaryDirectory() as tmpdir:
+            rfs_fname = f"{tmpdir}/rfs.json"
+            with open(rfs_fname, 'w') as f:
+                json.dump(rfs, f, indent=2, sort_keys=True)
```

Comment:
Should this and https://github.com/NeurodataWithoutBorders/lindi/blob/bc71a9d2035a5331091fc11821514ef9bccdef9d/lindi/LindiH5ZarrStore/LindiH5ZarrStore.py#L463 use the same function so that we have consistent file name suffixes and `json.dump` args?

Reply:
Done.